pkg/apis/core/validation/validation.go

```diff
@@ -4106,12 +4149,48 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 		return allErrs
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
+		// reject attempts to change pod qos
+		oldQoS := qos.GetPodQOS(oldPod)
```

Comment:
I don't get it.  If I started with limit > request, and then change it to limit == request - this will fail?  I'm not saying that the QOS should be mutable, but it shouldn't be invalid to lower your limit.

Reply:
In the specific case where we lower the limit to appear 'Guaranteed', yes this will fail. I know this is not ideal there's a bunch of code in kubelet (preemption, policy, pod status reporting) that use GetPodQOS to classify the pod and all this code didn't see the case where resources would become mutable.

If we allowed this, then a pod could come up as burstable pod cgroup, and slide over to Guaranteed class for preemption accounting.

Can we take an issue on this and attempt to fix in Beta?

Reply:
I reviewed all places where GetPodQOS is currently used - mostly kubelet and some of it makes assumptions that it won't change, and a few places in the API. On new pod create, it is called to initialize PodStatus.QOSClass. 

This is another of those places where it would be super-nice if we could rely on Status persistence and read from PodStatus.QOSClass once it is set instead of the unwieldy computation each time it is called.

I'll discuss with SIG-node if we can internally switch to getting QoS from cgroup before Beta. This would be a separate change/PR - I don't want to conflate it with my code.

Reply:
Please drop a big TODO comment here that explains this and why it exists and that it MUST be fixed for beta.

Using `status` to store the QoS should be acceptable under the approved-in-all-but-name KEP to loosen status (https://github.com/kubernetes/website/pull/28454)